### DISCUSSION TITLE
Add parameter in model_config for discarding ValidationError in optional fields

### DISCUSSION THREAD

**Discussion by ZehavaG** [pydantic/pydantic#8579] [Question] [answered]
hey!
I would love to know if there is any chance to get this cool feature in one of the upcoming releases of pydantic:
```py 
class ClassName(BaseModel):
    model_config = ConfigDict(use_default_when_validation_error=True)

    required_field: int
    optional_field: Optional[int] = None
```
I added the parameter "**use_default_when_validation_error**=True".
So in this case when I make an instance as following:
```py 
my_class = ClassName(required_field=3, optional_field="abc")
```
I would expect that my_class is created without error and with optional_field = None.
Meaning if there are fields that are optional, instead of getting a ValidationError when inputting a wrong type, use the default.

I found a similar request in this discussion: #2741, but there was no mention of including it in one of the upcoming pydantic versions, and the solution there is not relevant for pydantic v2.

Thank you.

**Comment by uriyyo** [comment, marked as answer]
Hi @ZehavaG,

You can define a base class that can provide features as you described:
```py
from functools import partial
from typing import Any, Callable, Optional

from pydantic import BaseModel, WrapValidator


def _fallback_to_default_validator(
    get_default: Callable[[], Any],
    v: Any,
    next_: Callable[[Any], Any],
) -> Any:
    try:
        return next_(v)
    except ValueError:
        return get_default()


class UseDefaultWhenValidationError(BaseModel):
    @classmethod
    def __pydantic_init_subclass__(cls, **kwargs: Any) -> None:
        for field in cls.model_fields.values():
            if not field.is_required():
                validator = WrapValidator(partial(_fallback_to_default_validator, field.get_default))
                field.metadata.append(validator)

        cls.model_rebuild(force=True)


class ClassName(UseDefaultWhenValidationError, BaseModel):
    required_field: int
    optional_field: Optional[int] = None


my_class = ClassName(required_field=3, optional_field="abc")

print(my_class)
```
```py
required_field=3 optional_field=None
```

**Reply by ZehavaG** [reply]
Thank you so much!

**Reply by uriyyo** [reply]
Happy to hear)
You can mark my answer as correct so others will now that this discussion has a solution.
